### PR TITLE
Allow overriding Task methods without @Internal annotation

### PR DESCRIPTION
**Pull request by asodja**
Try to fix https://github.com/gradle/gradle/issues/17500.

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy [hunk 1]
```diff
@@ -633,4 +633,75 @@ class TaskDefinitionIntegrationTest extends AbstractIntegrationSpec {
         then:
         failure.assertHasCause("Adding a task provider directly to the task container is not supported.")
     }
+
+    def "can override task name, description and group"() {
```
Comment (asodja) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy:637]: Note to myself: Also include `without @Internal annotation`

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java [hunk 2]
```diff
@@ -774,7 +772,6 @@ public String determineName(Task c) {
      * @since 5.0
      */
     @Internal
-    @Optional
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/Task.java:777]: Why shouldn't this be `@Optional`? We might not fail without the annotation, but the value _is_ _nullable_ (which `@Optional` actually means). We use `null` to express "no timeout."

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy [hunk 3]
```diff
@@ -633,4 +633,75 @@ class TaskDefinitionIntegrationTest extends AbstractIntegrationSpec {
...
+        when:
+        succeeds("customTask")
+
+        then:
+        outputContains("Hello from CustomTask")
+    }
+
+    // This test just shows the problem with overriding the task name,
+    // if it does not match name at the registration it will not be found.
+    def "fails when running a task with overriden name that is registered with a different name"() {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy:679]: I don't think we should allow overriding the task name. We should instead make that method `final` in 8.0, and just ignore the problem until them.

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java (hunk 2 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java:777]: The problem is, that with `@Optional` with this change I get a report that this field is `@Internal` and `@Optional` which is not allowed. Is `@Internal` annotation wrong here?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy (hunk 3 above)
Comment (asodja) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy:679]: Ok, I can remove this and I will open an issue to make it final in 8.0.

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java (hunk 2 above)
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java:777]: You are right, I missed the `@Internal` part. With that you don't need `@Optional` indeed.

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy (hunk 3 above)
Comment (asodja) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDefinitionIntegrationTest.groovy]: Removed this test.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by asodja** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
There's a network problem. I created https://github.com/gradle/gradle-private/issues/3504 to track it.

@bot-gradle test and merge

**Pull request merged**